src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -140,24 +161,35 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
             IStreamConsumerExtension streamConsumer, 
             IStreamFilterPredicateWrapper filter)
         {
+            counterConsumersAdded.Increment();
             PubSubSubscriptionState pubSubState = State.Consumers.FirstOrDefault(s => s.Equals(subscriptionId));
-            if (pubSubState == null)
+            try
             {
-                pubSubState = new PubSubSubscriptionState(subscriptionId, streamId, streamConsumer);
-                State.Consumers.Add(pubSubState);
-            }
+                if (pubSubState == null)
+                {
+                    pubSubState = new PubSubSubscriptionState(subscriptionId, streamId, streamConsumer);
+                    State.Consumers.Add(pubSubState);
+                }
 
-            if (pubSubState.IsFaulted)
-                throw new FaultedSubscriptionException(subscriptionId, streamId);
+                if (pubSubState.IsFaulted)
+                    throw new FaultedSubscriptionException(subscriptionId, streamId);
```

Comment:
if this is frequent enough, it would be worth moving up outside the try/catch (and before the state could be mutated). This way the grain is not unnecessarily deactivated.